wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/spi/jpa/JpaMemberDao.java

```diff
@@ -0,0 +1,219 @@
+    public List<OrganizationImpl> getOrganizations(String userId) throws ServerException {
+        requireNonNull(userId, "Required non-null user id");
+        try {
+            final EntityManager manager = managerProvider.get();
+            return manager.createNamedQuery("Member.getOrganizations", OrganizationImpl.class)
+                          .setParameter("userId", userId)
+                          .getResultList();
+        } catch (RuntimeException e) {
+            throw new ServerException(e.getLocalizedMessage(), e);
+        }
+    }
+
+    @Transactional
+    protected void doRemove(String organizationId, String userId) {
+        final EntityManager manager = managerProvider.get();
+        try {
+            MemberImpl member = manager.createNamedQuery("Member.getMember", MemberImpl.class)
+                                       .setParameter("userId", userId)
+                                       .setParameter("organizationId", organizationId)
+                                       .getSingleResult();
+            manager.remove(member);
+        } catch (NoResultException ignored) {
+            //member is already removed
+        }
+    }
+
+    @Singleton
+    public static class RemoveMembersBeforeOrganizationRemovedEventSubscriber implements EventSubscriber<BeforeOrganizationRemovedEvent> {
+        @Inject
+        private EventService eventService;
+        @Inject
+        private MemberDao    memberDao;
+
+        @PostConstruct
+        public void subscribe() {
+            eventService.subscribe(this);
+        }
+
+        @PreDestroy
+        public void unsubscribe() {
+            eventService.unsubscribe(this);
+        }
+
+        @Override
+        public void onEvent(BeforeOrganizationRemovedEvent event) {
+            try {
+                for (MemberImpl member : memberDao.getMembers(event.getOrganization().getId())) {
+                    memberDao.remove(member.getOrganizationId(), member.getUserId());
+                }
+            } catch (Exception x) {
+                LOG.error(format("Couldn't remove members before organization '%s' is removed", event.getOrganization().getId()), x);
+            }
+        }
+    }
+
+    @Singleton
+    public static class RemoveMembersBeforeUserRemovedEventSubscriber implements EventSubscriber<BeforeUserRemovedEvent> {
+        @Inject
+        private EventService eventService;
+        @Inject
+        private MemberDao    memberDao;
+
+        @PostConstruct
+        public void subscribe() {
+            eventService.subscribe(this);
+        }
+
+        @PreDestroy
+        public void unsubscribe() {
+            eventService.unsubscribe(this);
+        }
+
+        @Override
+        public void onEvent(BeforeUserRemovedEvent event) {
+            try {
+                for (MemberImpl member : memberDao.getMemberships(event.getUser().getId())) {
+                    memberDao.remove(member.getOrganizationId(), member.getUserId());
```

Comment:
Do we need to remove organization itself in some cases (when no more members with setPermissions/manage* roles left or etc)?